File: library/src/main/java/me/shkschneider/skeleton/ui/ViewHelper.java
package me.shkschneider.skeleton.ui;

import android.app.Activity;
import android.support.annotation.NonNull;
import android.view.View;
import android.view.ViewGroup;
import android.view.Window;

import java.util.ArrayList;
import java.util.List;

public class ViewHelper {

    protected ViewHelper() {
        // Empty
    }

    final static String ANDROIDXML = "http://schemas.android.com/apk/res/android";
    final static String RESAUTOXML = "http://schemas.android.com/apk/res-auto";

    public static <T extends View> T content(@NonNull final Activity activity) {
        return activity.findViewById(Window.ID_ANDROID_CONTENT);
    }

    public static List<View> children(@NonNull final View view) {
        final List<View> views = new ArrayList<>();
        final ViewGroup viewGroup = (ViewGroup) view;
        for (int i = 0; i < viewGroup.getChildCount(); i++) {
            final View v = viewGroup.getChildAt(i);
            views.add(v);
        }
        return views;
    }

    public static boolean show(@NonNull final View view) {
        if (view.getVisibility() != View.VISIBLE) {
            view.setVisibility(View.VISIBLE);
            return true;
        }
        return false;
    }

    public static boolean hide(@NonNull final View view) {
        if (view.getVisibility() != View.GONE) {
            view.setVisibility(View.GONE);
            return true;
        }
        return false;
    }

}

File: library/src/main/java/me/shkschneider/skeleton/ui/ViewHelper.kt
package me.shkschneider.skeleton.ui

import android.app.Activity
import android.view.View
import android.view.ViewGroup
import android.view.Window
import me.shkschneider.skeleton.SkeletonReceiver
import me.shkschneider.skeleton.java.SkHide
import java.util.*

object ViewHelper {

    val ANDROIDXML = "http://schemas.android.com/apk/res/android"
    val RESAUTOXML = "http://schemas.android.com/apk/res-auto"
    @SkHide
    val CONTENT = Window.ID_ANDROID_CONTENT
    @SkHide
    val NO_ID = View.NO_ID
    val RESULT_VISIBILITY = "VISIBILITY"

    fun <T : View> content(activity: Activity): T {
        return activity.findViewById(CONTENT)
    }

    fun children(view: View): List<View> {
        val viewGroup = view as ViewGroup
        return (0 until viewGroup.childCount).mapTo(ArrayList<View>()) { viewGroup.getChildAt(it) }
    }

    fun show(view: View): Boolean {
        if (view.visibility != View.VISIBLE) {
            view.visibility = View.VISIBLE
            return true
        }
        return false
    }

    fun hide(view: View): Boolean {
        if (view.visibility != View.GONE) {
            view.visibility = View.GONE
            return true
        }
        return false
    }

    // <https://stackoverflow.com/a/32778292>
    fun visibilityListener(view: View, skeletonReceiver: SkeletonReceiver) {
        view.tag = view.visibility
        view.viewTreeObserver.addOnGlobalLayoutListener {
            if (view.tag != view.visibility) {
                view.tag = view.visibility
                skeletonReceiver.post(RESULT_VISIBILITY, view.visibility)
            }
        }
    }

}
